<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="An interactive scientific data analysis pipeline for mass
    cytometry (CyTOF) data built using tidy data principles.  It is
    specifically designed to play well with the tidyverse suite of data
    cleaning and analysis tools, with functionality for input/output of
    mass cytometry files, data cleaning, preprocessing, clustering,
    visualization, modeling, and quality-of-life functions.">
<title>Analyze Mass Cytometry Data Using Tidy Data Principles • tidytof</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Analyze Mass Cytometry Data Using Tidy Data Principles">
<meta property="og:description" content="An interactive scientific data analysis pipeline for mass
    cytometry (CyTOF) data built using tidy data principles.  It is
    specifically designed to play well with the tidyverse suite of data
    cleaning and analysis tools, with functionality for input/output of
    mass cytometry files, data cleaning, preprocessing, clustering,
    visualization, modeling, and quality-of-life functions.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">tidytof</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="tidytof-an-ecosystem-for-tidy-and-highly-reproducible-cytof-data-analysis">tidytof: An ecosystem for tidy and highly-reproducible CyTOF data analysis<a class="anchor" aria-label="anchor" href="#tidytof-an-ecosystem-for-tidy-and-highly-reproducible-cytof-data-analysis"></a>
</h1></div>
<!-- badges: start -->

<!-- badges: end -->
<p>The goal of <code>tidytof</code> is to provide an integrated suite of R functions for CyTOF data analysis using an intuitive and internally consistent set of design principles. As an extension of the <code>tidyverse</code> ecosystem of data manipulation tools in R, <code>tidytof</code> is both performant and easy-to-use for scientists with a wide range of coding experience (including beginners).</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the development version of tidytof from GitHub with the following command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span> 
<span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"keyes-timothy/tidytof"</span><span class="op">)</span></code></pre></div>
<p>Once <code>{tidytof}</code> is installed, you can attach it to your current R session using the following code:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">tidytof</span><span class="op">)</span></code></pre></div>
<p>In addition, we can load the other packages we need for this vignette:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<div class="section level3">
<h3 id="reading-data-including-example-datasets">Reading data (including example datasets)<a class="anchor" aria-label="anchor" href="#reading-data-including-example-datasets"></a>
</h3>
<p><code>{tidytof}</code> comes bundled with several example mass cytometry datasets. To access the raw .fcs and .csv files containing these data, use the <code>tidytof_example_data</code> function. When called with no arguments, <code>tidytof_example_data</code> will return a character vector naming the datasets contained in {tidytof}:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/tidytof_example_data.html">tidytof_example_data</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; [1] "aml"                  "ddpr"                 "mix"                 </span>
<span class="co">#&gt; [4] "mix2"                 "phenograph"           "phenograph_csv"      </span>
<span class="co">#&gt; [7] "scaffold"             "statistical_scaffold" "surgery"</span></code></pre></div>
<p>To obtain the file path for the directory containing each dataset, call <code>tidytof_example_data</code> with one of these dataset names as its argument. For example, to obtain the directory for the phenograph data, we would use the following command:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/tidytof_example_data.html">tidytof_example_data</a></span><span class="op">(</span><span class="st">"phenograph"</span><span class="op">)</span>
<span class="co">#&gt; [1] "/Library/Frameworks/R.framework/Versions/4.1/Resources/library/tidytof/extdata/phenograph"</span></code></pre></div>
<p>Using one of these directories (or any other directory containing CyTOF data on your local machine), we can use <code>tof_read_data</code> to read CyTOF data from raw files. <code>tof_read_data</code> is smart enough to read single .fcs/.csv files or multiple .fcs/.csv files depending on whether its first argument (<code>path</code>) leads to a single file or to a directory of files.</p>
<p>Here, we can use <code>tof_read_data</code> to read in all of the .fcs files in the “phenograph” example dataset bundled into <code>{tidytof}</code> and store it in the <code>phenograph</code> variable.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">phenograph</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="reference/tidytof_example_data.html">tidytof_example_data</a></span><span class="op">(</span><span class="st">"phenograph"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tof_read_data.html">tof_read_data</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">phenograph</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 × 45</span>
<span class="co">#&gt;   file_name   `Time|Time` `Cell_length|Ce… `DNA1|Ir191` `DNA2|Ir193` `BC1|Pd104`</span>
<span class="co">#&gt;   &lt;chr&gt;             &lt;dbl&gt;            &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;</span>
<span class="co">#&gt; 1 H1_PhenoGr…         462               24         338.         663.        47.4</span>
<span class="co">#&gt; 2 H1_PhenoGr…        5856               29         277.         577.       214. </span>
<span class="co">#&gt; 3 H1_PhenoGr…        7621               25         289.         679.       223. </span>
<span class="co">#&gt; 4 H1_PhenoGr…        8783               24         215.         393.        62.8</span>
<span class="co">#&gt; 5 H1_PhenoGr…       10106               29         135.         342.        25.6</span>
<span class="co">#&gt; 6 H1_PhenoGr…       11326               22         181.         393.        34.5</span>
<span class="co">#&gt; # … with 39 more variables: BC2|Pd106 &lt;dbl&gt;, BC3|Pd108 &lt;dbl&gt;, BC4|In113 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   BC5|In115 &lt;dbl&gt;, BC6|La139 &lt;dbl&gt;, pPLCg2|Pr141 &lt;dbl&gt;, CD19|Nd142 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   p4EBP1|Nd143 &lt;dbl&gt;, CD11b|Nd144 &lt;dbl&gt;, pAMPK|Nd145 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   pSTAT3|Nd146 &lt;dbl&gt;, CD34|Nd148 &lt;dbl&gt;, pSTAT5|Nd150 &lt;dbl&gt;, pS6|Sm147 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   pCREB|Sm149 &lt;dbl&gt;, pc-Cbl|Sm152 &lt;dbl&gt;, CD45|Sm154 &lt;dbl&gt;, CD123|Eu151 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   pSTAT1|Eu153 &lt;dbl&gt;, pZap70-Syk|Gd156 &lt;dbl&gt;, CD33|Gd158 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   CD47|Gd160 &lt;dbl&gt;, pAKT|Tb159 &lt;dbl&gt;, CD7|Dy162 &lt;dbl&gt;, CD15|Dy164 &lt;dbl&gt;, …</span></code></pre></div>
<p>Regardless of its input format, <code>{tidytof}</code> reads data into an extended <code>tibble</code> called a <code>tof_tbl</code>, an S3 class identical to <code>tbl_df</code>, but with one additional attribute (“panel”). <code>{tidytof}</code> stores this additional attribute in <code>tof_tbl</code>s because, in addition to analyzing CyTOF data from individual experiments, CyTOF users often want to compare panels between experiments to find common markers or to compare which metals are associated with particular markers across panels.</p>
<p>A few notes about <code>tof_tbl</code>s:</p>
<ul>
<li>
<code>tof_tbl</code>s contains one cell per row and one channel per column (to provide the data in its “tidy” format).</li>
<li>
<code>tof_read_data</code> adds an additional column to the output <code>tof_tbl</code> encoding the name of the file from which each cell was read (the “file_name” column).</li>
<li>Because <code>tof_tbl</code>s inherit from the <code>tbl_df</code> class, all methods available to tibbles are also available to <code>tof_tbl</code>s. For example, <a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a>’s useful <code>mutate</code> method can be applied to our <code>tof_tbl</code> named <code>phenograph</code> above to convert the columns encoding the phenograph cluster and stimulation condition to which each cell belongs into character vectors (instead of their original numeric codes in the uncleaned dataset).</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">phenograph</span> <span class="op">&lt;-</span> 
  <span class="va">phenograph</span> <span class="op">%&gt;%</span> 
  <span class="co"># mutate the input tof_tbl</span>
  <span class="fu">mutate</span><span class="op">(</span>
    `PhenoGraph|PhenoGraph` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">`PhenoGraph|PhenoGraph`</span><span class="op">)</span>, 
    `Condition|Condition` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">`Condition|Condition`</span><span class="op">)</span>
  <span class="op">)</span>

<span class="va">phenograph</span> <span class="op">%&gt;%</span> 
  <span class="co"># use dplyr's select method to show that the columns have been changed</span>
  <span class="fu">select</span><span class="op">(</span><span class="fu">where</span><span class="op">(</span><span class="va">is.character</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 77,563 × 3</span>
<span class="co">#&gt;    file_name              `PhenoGraph|PhenoGraph` `Condition|Condition`</span>
<span class="co">#&gt;    &lt;chr&gt;                  &lt;chr&gt;                   &lt;chr&gt;                </span>
<span class="co">#&gt;  1 H1_PhenoGraph_cluster1 1                       1                    </span>
<span class="co">#&gt;  2 H1_PhenoGraph_cluster1 1                       1                    </span>
<span class="co">#&gt;  3 H1_PhenoGraph_cluster1 1                       1                    </span>
<span class="co">#&gt;  4 H1_PhenoGraph_cluster1 1                       1                    </span>
<span class="co">#&gt;  5 H1_PhenoGraph_cluster1 1                       1                    </span>
<span class="co">#&gt;  6 H1_PhenoGraph_cluster1 1                       1                    </span>
<span class="co">#&gt;  7 H1_PhenoGraph_cluster1 1                       1                    </span>
<span class="co">#&gt;  8 H1_PhenoGraph_cluster1 1                       1                    </span>
<span class="co">#&gt;  9 H1_PhenoGraph_cluster1 1                       1                    </span>
<span class="co">#&gt; 10 H1_PhenoGraph_cluster1 1                       1                    </span>
<span class="co">#&gt; # … with 77,553 more rows</span></code></pre></div>
<p>The <code>tof_tbl</code> class is preserved even after these transformations.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">phenograph</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; [1] "tof_tbl"    "tbl_df"     "tbl"        "data.frame"</span></code></pre></div>
<p>Finally, to retrieve panel information from a <code>tof_tbl</code>, use <code>tof_get_panel</code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">phenograph</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tof_get_panel.html">tof_get_panel</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 44 × 2</span>
<span class="co">#&gt;    metals      antigens   </span>
<span class="co">#&gt;    &lt;chr&gt;       &lt;chr&gt;      </span>
<span class="co">#&gt;  1 Time        Time       </span>
<span class="co">#&gt;  2 Cell_length Cell_length</span>
<span class="co">#&gt;  3 Ir191       DNA1       </span>
<span class="co">#&gt;  4 Ir193       DNA2       </span>
<span class="co">#&gt;  5 Pd104       BC1        </span>
<span class="co">#&gt;  6 Pd106       BC2        </span>
<span class="co">#&gt;  7 Pd108       BC3        </span>
<span class="co">#&gt;  8 In113       BC4        </span>
<span class="co">#&gt;  9 In115       BC5        </span>
<span class="co">#&gt; 10 La139       BC6        </span>
<span class="co">#&gt; # … with 34 more rows</span></code></pre></div>
<p>Importantly, <code>tof_read_data</code> uses an opinionated heuristic to mine different keyword slots of input .fcs file(s) and guess which metals and antigens were used during data collection. Thus, when .csv files are being read using <code>tof_read_data</code>, it is recommended to use the <code>panel_info</code> argument to provide the panel manually (as .csv files, unlike .fcs files, do not provide built-in metadata about the columns they contain).</p>
</div>
<div class="section level3">
<h3 id="pre-processing">Pre-processing<a class="anchor" aria-label="anchor" href="#pre-processing"></a>
</h3>
<p>Generally, the raw ion counts for each analyte measured on a mass cytometer need to be transformed before CyTOF data analysis. Common preprocessing steps may include variance-stabilizing transformations - such as the hyperbolic arcsine (arcsinh) transformation or a log transformation - scaling/centering, and/or denoising.</p>
<p>To perform standard preprocessing tasks with <code>{tidytof}</code>, use <code>tof_preprocess</code>. <code>tof_preprocess</code>’s default behavior is to apply the arcsinh transformation (with a cofactor of 5) to each numeric column in the input <code>tof_tibble</code> as well as to remove the gaussian noise that Fluidigm software adds to each ion count (this noise is added for visualization purposes, but for most analyses, removing it is recommended).</p>
<p>As an example, we can preprocess our <code>phenograph</code> <code>tof_tibble</code> above and see how our first few measurements change before and after.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># before preprocessing</span>
<span class="va">phenograph</span> <span class="op">%&gt;%</span> 
  <span class="fu">select</span><span class="op">(</span><span class="va">`CD45|Sm154`</span>, <span class="va">`CD34|Nd148`</span>, <span class="va">`CD38|Er167`</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 × 3</span>
<span class="co">#&gt;   `CD45|Sm154` `CD34|Nd148` `CD38|Er167`</span>
<span class="co">#&gt;          &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;</span>
<span class="co">#&gt; 1        857.         4.76         10.7 </span>
<span class="co">#&gt; 2         71.9       -0.392        23.2 </span>
<span class="co">#&gt; 3        434.         4.33          5.70</span>
<span class="co">#&gt; 4         70.4        1.19          7.96</span>
<span class="co">#&gt; 5        356.         0.807        11.2 </span>
<span class="co">#&gt; 6        146.         0.145        44.3</span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># perform preprocessing</span>
<span class="va">phenograph</span> <span class="op">&lt;-</span> 
  <span class="va">phenograph</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tof_preprocess.html">tof_preprocess</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># visualize new values</span>
<span class="va">phenograph</span> <span class="op">%&gt;%</span> 
  <span class="fu">select</span><span class="op">(</span><span class="va">`CD45|Sm154`</span>, <span class="va">`CD34|Nd148`</span>, <span class="va">`CD38|Er167`</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 × 3</span>
<span class="co">#&gt;   `CD45|Sm154` `CD34|Nd148` `CD38|Er167`</span>
<span class="co">#&gt;          &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;</span>
<span class="co">#&gt; 1         5.84        0.881         1.53</span>
<span class="co">#&gt; 2         3.36        0             2.27</span>
<span class="co">#&gt; 3         5.16        0.881         1.02</span>
<span class="co">#&gt; 4         3.35        0.390         1.25</span>
<span class="co">#&gt; 5         4.96        0.199         1.61</span>
<span class="co">#&gt; 6         4.07        0.199         2.89</span></code></pre></div>
<p>To alter <code>tof_preprocess</code>’s default behavior, change the <code>channel_cols</code> argument (to specify which columns of <code>tof_tibble</code> should be transformed) and the <code>transform_fun</code> argument (to specify which vector-valued function should be used to transform each of the <code>channel_cols</code>). To keep the gaussian noise added by Fluidigm software (or if you are working with a dataset that does not have this noise), set the <code>undo_noise</code> argument to <code>FALSE</code>.</p>
</div>
<div class="section level3">
<h3 id="downsampling">Downsampling<a class="anchor" aria-label="anchor" href="#downsampling"></a>
</h3>
<p>Often, CyTOF experiments collect tens or hundreds or millions of cells in total, and it can be useful to downsample to a smaller, more computationally tractable number of cells - either for a final analysis or while developing code. To do this, <code>{tidytof}</code> implements the <code>tof_downsample</code> verb, which allows downsampling using 3 methods.</p>
<p>Using <code>{tidytof}</code>’s built-in dataset <code>phenograph_data</code> (which is a smaller version of the dataset we read in ourselves above), we can see that the original size of the dataset is 2000 cells per cluster, or 6000 cells in total:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">phenograph_data</span> <span class="op">%&gt;%</span> 
  <span class="fu">count</span><span class="op">(</span><span class="va">phenograph_cluster</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 3 × 2</span>
<span class="co">#&gt;   phenograph_cluster     n</span>
<span class="co">#&gt;   &lt;chr&gt;              &lt;int&gt;</span>
<span class="co">#&gt; 1 cluster1            2000</span>
<span class="co">#&gt; 2 cluster2            2000</span>
<span class="co">#&gt; 3 cluster3            2000</span></code></pre></div>
<p>To randomly sample 200 cells per cluster, we can use <code>tof_downsample</code> using the “constant” <code>method</code>:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">phenograph_data</span> <span class="op">%&gt;%</span> 
  <span class="co"># downsample </span>
  <span class="fu"><a href="reference/tof_downsample.html">tof_downsample</a></span><span class="op">(</span>
    method <span class="op">=</span> <span class="st">"constant"</span>, 
    group_cols <span class="op">=</span> <span class="va">phenograph_cluster</span>, 
    num_cells <span class="op">=</span> <span class="fl">200</span>
  <span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># count the number of downsampled cells in each cluster</span>
  <span class="fu">count</span><span class="op">(</span><span class="va">phenograph_cluster</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 3 × 2</span>
<span class="co">#&gt;   phenograph_cluster     n</span>
<span class="co">#&gt;   &lt;chr&gt;              &lt;int&gt;</span>
<span class="co">#&gt; 1 cluster1             200</span>
<span class="co">#&gt; 2 cluster2             200</span>
<span class="co">#&gt; 3 cluster3             200</span></code></pre></div>
<p>Alternatively, if we wanted to sample 50% of the cells in each cluster, we could use the “prop” <code>method</code>:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">phenograph_data</span> <span class="op">%&gt;%</span> 
  <span class="co"># downsample</span>
  <span class="fu"><a href="reference/tof_downsample.html">tof_downsample</a></span><span class="op">(</span>
    method <span class="op">=</span> <span class="st">"prop"</span>,
    group_cols <span class="op">=</span> <span class="va">phenograph_cluster</span>, 
    prop_cells <span class="op">=</span> <span class="fl">0.5</span>
  <span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># count the number of downsampled cells in each cluster</span>
  <span class="fu">count</span><span class="op">(</span><span class="va">phenograph_cluster</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 3 × 2</span>
<span class="co">#&gt;   phenograph_cluster     n</span>
<span class="co">#&gt;   &lt;chr&gt;              &lt;int&gt;</span>
<span class="co">#&gt; 1 cluster1            1000</span>
<span class="co">#&gt; 2 cluster2            1000</span>
<span class="co">#&gt; 3 cluster3            1000</span></code></pre></div>
<p>Alternatively, you might be interested in taking a slightly different approach to downsampling that downsamples the number of cells not to a fixed constant or proportion, but to a fixed <em>density</em> in phenotypic space. For example, the following scatterplot demonstrates that there are certain areas of phenotypic density in <code>phenograph_data</code> that contain more cells than others along the cd34/cd38 axes:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">phenograph_data</span> <span class="op">%&gt;%</span> 
  <span class="co"># preprocess all numeric columns in the dataset</span>
  <span class="fu"><a href="reference/tof_preprocess.html">tof_preprocess</a></span><span class="op">(</span>undo_noise <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># make a scatterplot</span>
  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">cd34</span>, y <span class="op">=</span> <span class="va">cd38</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">geom_point</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">scale_x_continuous</span><span class="op">(</span>limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="fl">1.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">scale_y_continuous</span><span class="op">(</span>limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-16-1.png" width="100%"></p>
<p>To reduce the number of cells in our dataset until the local density around each cell in our dataset is relatively constant, we can use the “density” <code>method</code> of <code>tof_downsample</code>:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">phenograph_data</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tof_preprocess.html">tof_preprocess</a></span><span class="op">(</span>undo_noise <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tof_downsample.html">tof_downsample</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"density"</span>, density_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">cd34</span>, <span class="va">cd38</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">cd34</span>, y <span class="op">=</span> <span class="va">cd38</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">geom_point</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">scale_x_continuous</span><span class="op">(</span>limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="fl">1.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">scale_y_continuous</span><span class="op">(</span>limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-17-1.png" width="100%"></p>
<p>For more details, check out the documentation for the 3 underlying members of the <code>tof_downsample_*</code> function family (which are wrapped by <code>tof_downsample</code>):</p>
<ul>
<li><code>tof_downsample_constant</code></li>
<li><code>tof_downsample_prop</code></li>
<li><code>tof_downsample_density</code></li>
</ul>
</div>
<div class="section level3">
<h3 id="clustering">Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h3>
<p>Once input files are read into a tabular format and preprocessed/downsampled, we might be interested in clustering our data to define communities of cells with shared characteristics.</p>
<p>To do so, we can use the <code>tof_cluster</code> verb. Several clustering methods are implemented in <code>{tidytof}</code>, including <a href="https://pubmed.ncbi.nlm.nih.gov/25573116/" class="external-link">FlowSOM</a>, <a href="https://pubmed.ncbi.nlm.nih.gov/26095251/" class="external-link">PhenoGraph</a>, k-means, and others.</p>
<p>To demonstrate, we can apply the FlowSOM clustering algorithm to our <code>phenograph_data</code> from above. Note that <code>phenograph_data</code> contains 6000 total cells (2000 each from 3 clusters identified in the <a href="https://pubmed.ncbi.nlm.nih.gov/26095251/" class="external-link">original PhenoGraph publication</a>).</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">phenograph_clusters</span> <span class="op">&lt;-</span> 
  <span class="va">phenograph_data</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tof_preprocess.html">tof_preprocess</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tof_cluster.html">tof_cluster</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"flowsom"</span>, cluster_cols <span class="op">=</span> <span class="fu">contains</span><span class="op">(</span><span class="st">"cd"</span><span class="op">)</span><span class="op">)</span>

<span class="va">phenograph_clusters</span> <span class="op">%&gt;%</span> 
  <span class="fu">select</span><span class="op">(</span><span class="va">sample_name</span>, <span class="va">.flowsom_metacluster</span>, <span class="fu">everything</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 × 27</span>
<span class="co">#&gt;   sample_name   .flowsom_metaclu… phenograph_clus…  cd19 cd11b  cd34  cd45 cd123</span>
<span class="co">#&gt;   &lt;chr&gt;         &lt;chr&gt;             &lt;chr&gt;            &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1 H1_PhenoGrap… 1                 cluster1         0.199  2.19 0      4.56 0.569</span>
<span class="co">#&gt; 2 H1_PhenoGrap… 1                 cluster1         0      2.92 0      5.33 1.61 </span>
<span class="co">#&gt; 3 H1_PhenoGrap… 1                 cluster1         0      1.99 0.390  4.62 0    </span>
<span class="co">#&gt; 4 H1_PhenoGrap… 1                 cluster1         0.881  1.53 0.199  4.61 1.02 </span>
<span class="co">#&gt; 5 H1_PhenoGrap… 1                 cluster1         0.881  2.05 0.199  4.80 0.733</span>
<span class="co">#&gt; 6 H1_PhenoGrap… 1                 cluster1         0      1.25 0.733  5.01 0.733</span>
<span class="co">#&gt; # … with 19 more variables: cd33 &lt;dbl&gt;, cd47 &lt;dbl&gt;, cd7 &lt;dbl&gt;, cd15 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   cd44 &lt;dbl&gt;, cd38 &lt;dbl&gt;, cd3 &lt;dbl&gt;, cd117 &lt;dbl&gt;, cd64 &lt;dbl&gt;, cd41 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   pstat3 &lt;dbl&gt;, pstat5 &lt;dbl&gt;, pampk &lt;dbl&gt;, p4ebp1 &lt;dbl&gt;, ps6 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   pcreb &lt;dbl&gt;, pzap70-syk &lt;dbl&gt;, prb &lt;dbl&gt;, perk1-2 &lt;dbl&gt;</span></code></pre></div>
<p>The output of <code>tof_cluster</code> is a <code>tof_tbl</code> identical to the input tibble, now with the addition of an additional column (“.flowsom_metacluster”) that encodes the cluster id for each cell in the input <code>tof_tbl</code>. Note that all output columns added to a tibble or <code>tof_tbl</code> by <code>{tidytof}</code> begin with a full-stop (“.”) to reduce the likelihood of collisions with existing column names.</p>
<p>Because the output of <code>tof_cluster</code> is a <code>tof_tbl</code>, we can use the <code>count</code> method to assess the accuracy of the FlowSOM clustering compared to the original clustering from the PhenoGraph paper.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">phenograph_clusters</span> <span class="op">%&gt;%</span> 
  <span class="fu">count</span><span class="op">(</span><span class="va">phenograph_cluster</span>, <span class="va">.flowsom_metacluster</span>, sort <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 × 3</span>
<span class="co">#&gt;   phenograph_cluster .flowsom_metacluster     n</span>
<span class="co">#&gt;   &lt;chr&gt;              &lt;chr&gt;                &lt;int&gt;</span>
<span class="co">#&gt; 1 cluster1           1                     1997</span>
<span class="co">#&gt; 2 cluster2           3                     1973</span>
<span class="co">#&gt; 3 cluster3           2                     1955</span>
<span class="co">#&gt; 4 cluster3           1                       45</span>
<span class="co">#&gt; 5 cluster2           2                       27</span>
<span class="co">#&gt; 6 cluster1           2                        3</span></code></pre></div>
<p>Here, we can see that the FlowSOM algorithm groups most cells from the same PhenoGraph cluster with one another (with a small number of mistakes per PhenoGraph cluster).</p>
<p>To change which clustering algorithm <code>tof_cluster</code> uses, alter the <code>method</code> flag; to change the columns used to compute the clusters, change the <code>cluster_cols</code> flag. And finally, if you want to return a <code>tibble</code> that only includes the cluster labels (not the cluster labels added as a new column to the input <code>tof_tbl</code>), set <code>add_col</code> to <code>FALSE</code>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># will result in a tibble with only 1 column (the cluster labels)</span>
<span class="va">phenograph_data</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tof_preprocess.html">tof_preprocess</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tof_cluster.html">tof_cluster</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"flowsom"</span>, cluster_cols <span class="op">=</span> <span class="fu">contains</span><span class="op">(</span><span class="st">"cd"</span><span class="op">)</span>, add_col <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 × 1</span>
<span class="co">#&gt;   .flowsom_metacluster</span>
<span class="co">#&gt;   &lt;chr&gt;               </span>
<span class="co">#&gt; 1 3                   </span>
<span class="co">#&gt; 2 3                   </span>
<span class="co">#&gt; 3 3                   </span>
<span class="co">#&gt; 4 3                   </span>
<span class="co">#&gt; 5 3                   </span>
<span class="co">#&gt; 6 3</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="dimensionality-reduction">Dimensionality reduction<a class="anchor" aria-label="anchor" href="#dimensionality-reduction"></a>
</h3>
<p><code>{tidytof}</code> includes several algorithms for dimensionality reduction, a process whereby high-dimensional datasets can be represented in a smaller, easier-to-visualize number of dimensions. <code>{tidytof}</code> supports 3 dimensionality reduction methods: PCA, tSNE, and UMAP. To apply these to a dataset, use <code>tof_reduce_dimensions</code>:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># perform the dimensionality reduction</span>
<span class="va">phenograph_tsne</span> <span class="op">&lt;-</span> 
  <span class="va">phenograph_clusters</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tof_reduce_dimensions.html">tof_reduce_dimensions</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"tsne"</span><span class="op">)</span>

<span class="co"># select only the tsne embedding columns using a tidyselect helper (contains)</span>
<span class="va">phenograph_tsne</span> <span class="op">%&gt;%</span> 
  <span class="fu">select</span><span class="op">(</span><span class="fu">contains</span><span class="op">(</span><span class="st">"tsne"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6,000 × 2</span>
<span class="co">#&gt;    .tsne_1 .tsne_2</span>
<span class="co">#&gt;      &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt;  1   15.4    4.90 </span>
<span class="co">#&gt;  2   11.8    3.48 </span>
<span class="co">#&gt;  3   27.0   -5.30 </span>
<span class="co">#&gt;  4   10.4    0.450</span>
<span class="co">#&gt;  5   17.9    0.486</span>
<span class="co">#&gt;  6   17.4    1.41 </span>
<span class="co">#&gt;  7   17.2    6.73 </span>
<span class="co">#&gt;  8   28.4    0.970</span>
<span class="co">#&gt;  9    9.77  -0.401</span>
<span class="co">#&gt; 10   22.7  -12.9  </span>
<span class="co">#&gt; # … with 5,990 more rows</span></code></pre></div>
<p>By default, <code>tof_reduce_dimensions</code> will add reduced-dimension feature embeddings to the input <code>tof_tbl</code> and return the augmented <code>tof_tbl</code> as its result. To return only the features embeddings themselves, set <code>add_cols</code> to <code>FALSE</code> (as in <code>tof_cluster</code>).</p>
<p>Regardless of the method used, reduced-dimension feature embeddings can be used for visualization using <a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a> (or any graphics package):</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># plot the tsne embeddings using color to distinguish between clusters</span>
<span class="va">phenograph_tsne</span> <span class="op">%&gt;%</span> 
  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">.tsne_1</span>, y <span class="op">=</span> <span class="va">.tsne_2</span>, fill <span class="op">=</span> <span class="va">phenograph_cluster</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">geom_point</span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">21</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">labs</span><span class="op">(</span>fill <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-22-1.png" width="100%"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># plot the tsne embeddings using color to represent CD11b expression</span>
<span class="va">phenograph_tsne</span> <span class="op">%&gt;%</span> 
  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">.tsne_1</span>, y <span class="op">=</span> <span class="va">.tsne_2</span>, fill <span class="op">=</span> <span class="va">cd11b</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">geom_point</span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">21</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">scale_fill_viridis_c</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">labs</span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"CD11b expression"</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-22-2.png" width="100%"></p>
<p>Such visualizations can be helpful in qualitatively describing the phenotypic distinctions between the clusters present in a dataset.</p>
</div>
<div class="section level3">
<h3 id="differential-discovery-analysis">Differential Discovery Analysis<a class="anchor" aria-label="anchor" href="#differential-discovery-analysis"></a>
</h3>
<p>While dimensionality reduction can be used to visualize a clustering result, many CyTOF users also want to use statistical tools to rigorously quantify which clusters(s) in their dataset associated with a particular experimental or clinical variable.</p>
<p>Such analyses are often grouped under the umbrella term <strong>differential discovery analysis</strong> and include both comparing the relative <em>size</em> of clusters between experimental conditions (<strong>differential abundance analysis; DAA</strong>) as well as comparing marker expression patterns of clusters between experimental conditions (<strong>differential expression analysis; DEA</strong>). <code>{tidytof}</code> provides the <code>tof_daa</code> and <code>tof_dea</code> verbs for differential abundance and differential expression analyses, respectively.</p>
<p>To demonstrate how to use these verbs, we’ll first download a dataset originally collected for the development of the <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4084463/" class="external-link">CITRUS</a> algorithm. These data are available in the <a href="https://github.com/lmweber/HDCytoData" class="external-link">HDCytoData</a> package, which is available on Bioconductor and can be downloaded with the following command:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"HDCytoData"</span><span class="op">)</span></code></pre></div>
<p>To load the CITRUS data into our current R session, we can call a function from the <a href="https://github.com/lmweber/HDCytoData" class="external-link">HDCytoData</a>, which will provide it to us in a format from the <code>{flowCore}</code> package (called a “flowSet”). To convert this into a tidy tibble, we can use <code>{tidytof}</code> built-in method for converting flowCore objects into <code>tof_tbl</code>’s .</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">citrus_raw</span> <span class="op">&lt;-</span> <span class="fu">HDCytoData</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/HDCytoData/man/Bodenmiller_BCR_XL_SE.html" class="external-link">Bodenmiller_BCR_XL_flowSet</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">citrus_data</span> <span class="op">&lt;-</span> 
  <span class="va">citrus_raw</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/as_tof_tbl.html">as_tof_tbl</a></span><span class="op">(</span>sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">citrus_data</span><span class="op">)</span>
<span class="co">#&gt; [1] 172791     39</span></code></pre></div>
<p>Thus, we can see that <code>citrus_data</code> is a <code>tof_tbl</code> with 172791 cells (one in each row) and 39 pieces of information about each cell (one in each column).</p>
<p>We can also extract some metadata from the raw data and join it with our single-cell data using some <code>tidyverse</code> functions:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">citrus_metadata</span> <span class="op">&lt;-</span> 
  <span class="fu">tibble</span><span class="op">(</span>
    file_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu">flowCore</span><span class="fu">::</span><span class="fu">pData</span><span class="op">(</span><span class="va">citrus_raw</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, 
    sample_id <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">file_name</span><span class="op">)</span>,
    patient <span class="op">=</span> <span class="fu">str_extract</span><span class="op">(</span><span class="va">file_name</span>, <span class="st">"patient[:digit:]"</span><span class="op">)</span>, 
    stimulation <span class="op">=</span> <span class="fu">str_extract</span><span class="op">(</span><span class="va">file_name</span>, <span class="st">"(BCR-XL)|Reference"</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">mutate</span><span class="op">(</span>
    stimulation <span class="op">=</span> <span class="fu">if_else</span><span class="op">(</span><span class="va">stimulation</span> <span class="op">==</span> <span class="st">"Reference"</span>, <span class="st">"Basal"</span>, <span class="va">stimulation</span><span class="op">)</span>
  <span class="op">)</span>

<span class="va">citrus_metadata</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 × 4</span>
<span class="co">#&gt;   file_name                          sample_id patient  stimulation</span>
<span class="co">#&gt;   &lt;chr&gt;                                  &lt;int&gt; &lt;chr&gt;    &lt;chr&gt;      </span>
<span class="co">#&gt; 1 PBMC8_30min_patient1_BCR-XL.fcs            1 patient1 BCR-XL     </span>
<span class="co">#&gt; 2 PBMC8_30min_patient1_Reference.fcs         2 patient1 Basal      </span>
<span class="co">#&gt; 3 PBMC8_30min_patient2_BCR-XL.fcs            3 patient2 BCR-XL     </span>
<span class="co">#&gt; 4 PBMC8_30min_patient2_Reference.fcs         4 patient2 Basal      </span>
<span class="co">#&gt; 5 PBMC8_30min_patient3_BCR-XL.fcs            5 patient3 BCR-XL     </span>
<span class="co">#&gt; 6 PBMC8_30min_patient3_Reference.fcs         6 patient3 Basal</span></code></pre></div>
<p>Finally, we can join the metadata with our single-cell <code>tof_tbl</code> to obtain the cleaned dataset.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">citrus_data</span> <span class="op">&lt;-</span> 
  <span class="va">citrus_data</span> <span class="op">%&gt;%</span> 
  <span class="fu">left_join</span><span class="op">(</span><span class="va">citrus_metadata</span>, by <span class="op">=</span> <span class="st">"sample_id"</span><span class="op">)</span></code></pre></div>
<p>After data cleaning, we have <code>citrus_data</code>, a <code>tof_tbl</code> containing cells collected from 8 patients. Specifically, 2 samples were taken from each patient: one in which the cells’ B-cell receptors were stimulated (BCR-XL) and one in which they were not (Basal). In <code>citrus_data</code>, each cell’s patient of origin is stored in the <code>patient</code> column, and each cell’s stimulation condition is stored in the <code>stimulation</code> column. In addition, the <code>population_id</code> column stores information about cluster labels that were applied to each cell using a combination of FlowSOM clustering and manual merging (for details, run <code><a href="https://rdrr.io/pkg/HDCytoData/man/Bodenmiller_BCR_XL_SE.html" class="external-link">?HDCytoData::Bodenmiller_BCR_XL</a></code> in the R console).</p>
<p>We might be interested if there are certain clusters that expand or deplete within patients between the two stimulation conditions described above - this is a question of differential abundance analysis (DAA). <code>{tidytof}</code>’s <code>tof_daa</code> verb supports the use of 3 statistical approaches for performing DAA: diffcyt, generalized-linear mixed modeling (GLMMs), and simple t-tests. Because the setup described above uses a paired design and only has 2 experimental conditions of interest (Basal vs. BCR-XL), we can use the paired t-test method:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">daa_result</span> <span class="op">&lt;-</span> 
  <span class="va">citrus_data</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tof_daa.html">tof_daa</a></span><span class="op">(</span>
    daa_method <span class="op">=</span> <span class="st">"ttest"</span>, 
    cluster_col <span class="op">=</span> <span class="va">population_id</span>, 
    effect_col <span class="op">=</span> <span class="va">stimulation</span>, 
    group_cols <span class="op">=</span> <span class="va">patient</span>, 
    test_type <span class="op">=</span> <span class="st">"paired"</span>
  <span class="op">)</span>

<span class="va">daa_result</span>
<span class="co">#&gt; # A tibble: 8 × 8</span>
<span class="co">#&gt;   population_id    p_val   p_adj significant      t    df mean_diff mean_fc</span>
<span class="co">#&gt;   &lt;chr&gt;            &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;        &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt; 1 1             0.000924 0.00535 "*"         -5.48      7  -0.00743   0.644</span>
<span class="co">#&gt; 2 2             0.00623  0.0166  "*"         -3.86      7  -0.0156    0.674</span>
<span class="co">#&gt; 3 3             0.0235   0.0314  "*"         -2.88      7  -0.0638    0.849</span>
<span class="co">#&gt; 4 4             0.0235   0.0314  "*"          2.88      7   0.0832    1.38 </span>
<span class="co">#&gt; 5 5             0.0116   0.0232  "*"          3.39      7   0.00246   1.08 </span>
<span class="co">#&gt; 6 6             0.371    0.371   ""          -0.955     7  -0.0168    0.919</span>
<span class="co">#&gt; 7 7             0.00134  0.00535 "*"          5.14      7   0.0202    1.14 </span>
<span class="co">#&gt; 8 8             0.236    0.270   ""          -1.30      7  -0.00228   0.901</span></code></pre></div>
<p>Based on this output, we can see that 6 of our 8 clusters have statistically different abundance in our two stimulation conditions. Using <code>{tidytof}</code> easy integration with <a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a> packages, we can use this result to visualize the fold-changes of each cluster (within each patient) in the BCR-XL condition compared to the Basal condition using <a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a>:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plot_data</span> <span class="op">&lt;-</span> 
  <span class="va">citrus_data</span> <span class="op">%&gt;%</span> 
  <span class="fu">mutate</span><span class="op">(</span>population_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">population_id</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">left_join</span><span class="op">(</span>
    <span class="fu">select</span><span class="op">(</span><span class="va">daa_result</span>, <span class="va">population_id</span>, <span class="va">significant</span>, <span class="va">mean_fc</span><span class="op">)</span>, 
    by <span class="op">=</span> <span class="st">"population_id"</span>
  <span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">patient</span>, <span class="va">stimulation</span>, <span class="va">population_id</span>, <span class="va">significant</span>, <span class="va">mean_fc</span>, name <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">group_by</span><span class="op">(</span><span class="va">patient</span>, <span class="va">stimulation</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">mutate</span><span class="op">(</span>prop <span class="op">=</span> <span class="va">n</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">ungroup</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">pivot_wider</span><span class="op">(</span>
    names_from <span class="op">=</span> <span class="va">stimulation</span>, 
    values_from <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">prop</span>, <span class="va">n</span><span class="op">)</span>, 
  <span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">mutate</span><span class="op">(</span>
    diff <span class="op">=</span> <span class="va">`prop_BCR-XL`</span> <span class="op">-</span> <span class="va">prop_Basal</span>, 
    fc <span class="op">=</span> <span class="va">`prop_BCR-XL`</span> <span class="op">/</span> <span class="va">prop_Basal</span>,
    population_id <span class="op">=</span> <span class="fu">fct_reorder</span><span class="op">(</span><span class="va">population_id</span>, <span class="va">diff</span><span class="op">)</span>,
    direction <span class="op">=</span> 
      <span class="fu">case_when</span><span class="op">(</span>
        <span class="va">mean_fc</span> <span class="op">&gt;</span> <span class="fl">1</span> <span class="op">&amp;</span> <span class="va">significant</span> <span class="op">==</span> <span class="st">"*"</span> <span class="op">~</span> <span class="st">"increase"</span>,
        <span class="va">mean_fc</span> <span class="op">&lt;</span> <span class="fl">1</span> <span class="op">&amp;</span> <span class="va">significant</span> <span class="op">==</span> <span class="st">"*"</span> <span class="op">~</span> <span class="st">"decrease"</span>, 
        <span class="cn">TRUE</span> <span class="op">~</span> <span class="cn">NA_character_</span>
      <span class="op">)</span>
  <span class="op">)</span> 

<span class="va">significance_data</span> <span class="op">&lt;-</span> 
  <span class="va">plot_data</span> <span class="op">%&gt;%</span> 
  <span class="fu">group_by</span><span class="op">(</span><span class="va">population_id</span>, <span class="va">significant</span>, <span class="va">direction</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">summarize</span><span class="op">(</span>diff <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">diff</span><span class="op">)</span>, fc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">fc</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">ungroup</span><span class="op">(</span><span class="op">)</span>

<span class="va">significance_data</span>
<span class="co">#&gt; # A tibble: 8 × 5</span>
<span class="co">#&gt;   population_id significant direction     diff    fc</span>
<span class="co">#&gt;   &lt;fct&gt;         &lt;chr&gt;       &lt;chr&gt;        &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1 3             "*"         decrease  -0.00587 0.982</span>
<span class="co">#&gt; 2 2             "*"         decrease  -0.00526 0.882</span>
<span class="co">#&gt; 3 6             ""          &lt;NA&gt;       0.0676  1.80 </span>
<span class="co">#&gt; 4 1             "*"         decrease  -0.00288 0.799</span>
<span class="co">#&gt; 5 8             ""          &lt;NA&gt;       0.00571 1.42 </span>
<span class="co">#&gt; 6 5             "*"         increase   0.00520 1.60 </span>
<span class="co">#&gt; 7 7             "*"         increase   0.0386  1.27 </span>
<span class="co">#&gt; 8 4             "*"         increase   0.224   1.95</span>

<span class="va">plot_data</span> <span class="op">%&gt;%</span> 
  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">population_id</span>, y <span class="op">=</span> <span class="va">fc</span>, fill <span class="op">=</span> <span class="va">direction</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">geom_violin</span><span class="op">(</span>trim <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_hline</span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">1</span>, color <span class="op">=</span> <span class="st">"red"</span>, linetype <span class="op">=</span> <span class="st">"dotted"</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">geom_text</span><span class="op">(</span>
    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">population_id</span>, y <span class="op">=</span> <span class="va">fc</span>, label <span class="op">=</span> <span class="va">significant</span><span class="op">)</span>, 
    data <span class="op">=</span> <span class="va">significance_data</span>, 
    size <span class="op">=</span> <span class="fl">8</span>, 
    nudge_x <span class="op">=</span> <span class="fl">0.2</span>,
    nudge_y <span class="op">=</span> <span class="fl">0.06</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu">scale_x_discrete</span><span class="op">(</span>labels <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">str_c</span><span class="op">(</span><span class="st">"cluster "</span>, <span class="va">x</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">scale_fill_discrete</span><span class="op">(</span>na.translate <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">labs</span><span class="op">(</span>
    x <span class="op">=</span> <span class="cn">NULL</span>, 
    y <span class="op">=</span> <span class="st">"Abundance fold-chance (stimulated / basal)"</span>, 
    fill <span class="op">=</span> <span class="st">"Effect"</span>, 
    caption <span class="op">=</span> <span class="st">"Asterisks indicate significance at an adjusted p-value of 0.05"</span>
  <span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-28-1.png" width="100%"></p>
<p>Importantly, the output of <code>tof_daa</code> depends slightly on the underlying statistical method being used, and details can be found in the documentation for each <code>tof_daa_*</code> function family member:</p>
<ul>
<li><code>tof_daa_diffcyt</code></li>
<li><code>tof_daa_glmm</code></li>
<li><code>tof_daa_ttest</code></li>
</ul>
<p>Similarly, suppose we’re interested in how intracellular signaling proteins change their expression levels between our two stimulation conditions in each of our clusters. This is a Differential Expression Analysis (DEA) and can be performed using <code>{tidytof}</code>’s <code>tof_dea</code> verb. As above, we can use paired t-tests with multiple-hypothesis correction to to test for significant differences in each cluster’s expression of our signaling markers between stimulation conditions.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">signaling_markers</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>
    <span class="st">"pNFkB_Nd142"</span>, <span class="st">"pStat5_Nd150"</span>, <span class="st">"pAkt_Sm152"</span>, <span class="st">"pStat1_Eu153"</span>, <span class="st">"pStat3_Gd158"</span>, 
    <span class="st">"pSlp76_Dy164"</span>, <span class="st">"pBtk_Er166"</span>, <span class="st">"pErk_Er168"</span>, <span class="st">"pS6_Yb172"</span>, <span class="st">"pZap70_Gd156"</span>
  <span class="op">)</span>

<span class="va">dea_result</span> <span class="op">&lt;-</span> 
  <span class="va">citrus_data</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tof_preprocess.html">tof_preprocess</a></span><span class="op">(</span>channel_cols <span class="op">=</span> <span class="fu">any_of</span><span class="op">(</span><span class="va">signaling_markers</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tof_dea.html">tof_dea</a></span><span class="op">(</span>
    dea_method <span class="op">=</span> <span class="st">"ttest"</span>, 
    cluster_col <span class="op">=</span> <span class="va">population_id</span>, 
    marker_cols <span class="op">=</span> <span class="fu">any_of</span><span class="op">(</span><span class="va">signaling_markers</span><span class="op">)</span>, 
    effect_col <span class="op">=</span> <span class="va">stimulation</span>,
    group_cols <span class="op">=</span> <span class="va">patient</span>, 
    test_type <span class="op">=</span> <span class="st">"paired"</span>
  <span class="op">)</span>

<span class="va">dea_result</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 × 9</span>
<span class="co">#&gt;   population_id marker   p_val   p_adj significant     t    df mean_diff mean_fc</span>
<span class="co">#&gt;   &lt;chr&gt;         &lt;chr&gt;    &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;       &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt; 1 1             pS6_Y… 7.98e-8 2.36e-6 *            22.8     7     2.50    3.95 </span>
<span class="co">#&gt; 2 2             pS6_Y… 1.12e-7 2.36e-6 *            21.7     7     2.09    2.41 </span>
<span class="co">#&gt; 3 3             pBtk_… 1.47e-7 2.36e-6 *           -20.8     7    -0.459   0.383</span>
<span class="co">#&gt; 4 7             pBtk_… 1.28e-7 2.36e-6 *           -21.3     7    -0.500   0.374</span>
<span class="co">#&gt; 5 8             pBtk_… 1.10e-7 2.36e-6 *           -21.7     7    -0.500   0.403</span>
<span class="co">#&gt; 6 4             pBtk_… 8.24e-7 1.10e-5 *           -16.2     7    -0.447   0.392</span></code></pre></div>
<p>While the output of <code>tof_dea</code> also depends on the underlying test being used, we can see that the result above looks relatively similar to the output from <code>tof_daa</code>. Above, the output is a tibble in which each row represents the differential expression results from a single cluster-marker pair - for example, the first row represents the difference in expression of pS6 in cluster 1 between the BCR-XL and Basal conditions. Each row includes the raw p-value and multiple-hypothesis-corrected p-value for each cluster-marker pair.</p>
<p>This result can be used to make a volcano plot to visualize the results for all cluster-marker pairs:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">volcano_data</span> <span class="op">&lt;-</span> 
  <span class="va">dea_result</span> <span class="op">%&gt;%</span> 
  <span class="fu">mutate</span><span class="op">(</span>
    log2_fc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">mean_fc</span>, base <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>, 
    log_p <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">p_adj</span><span class="op">)</span>, 
    significance <span class="op">=</span> 
      <span class="fu">case_when</span><span class="op">(</span>
        <span class="va">p_adj</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">mean_fc</span> <span class="op">&gt;</span> <span class="fl">1</span> <span class="op">~</span> <span class="st">"increased"</span>, 
        <span class="va">p_adj</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">mean_fc</span> <span class="op">&lt;</span> <span class="fl">1</span> <span class="op">~</span> <span class="st">"decreased"</span>,
        <span class="cn">TRUE</span> <span class="op">~</span> <span class="cn">NA_character_</span>
      <span class="op">)</span>, 
    marker <span class="op">=</span> 
      <span class="fu">str_extract</span><span class="op">(</span><span class="va">marker</span>, <span class="st">".+_"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
      <span class="fu">str_remove</span><span class="op">(</span><span class="st">"_"</span><span class="op">)</span>, 
    pair <span class="op">=</span> <span class="fu">str_c</span><span class="op">(</span><span class="va">marker</span>, <span class="fu">str_c</span><span class="op">(</span><span class="st">"cluster "</span>, <span class="va">population_id</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"@"</span><span class="op">)</span>
  <span class="op">)</span>

<span class="va">volcano_data</span> <span class="op">%&gt;%</span> 
  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">log2_fc</span>, y <span class="op">=</span> <span class="va">log_p</span>, fill <span class="op">=</span> <span class="va">significance</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">geom_vline</span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, color <span class="op">=</span> <span class="st">"gray50"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">geom_hline</span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">geom_point</span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">21</span>, size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggrepel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html" class="external-link">geom_text_repel</a></span><span class="op">(</span>
    <span class="fu">aes</span><span class="op">(</span>label <span class="op">=</span> <span class="va">pair</span><span class="op">)</span>, 
    data <span class="op">=</span> <span class="fu">slice_head</span><span class="op">(</span><span class="va">volcano_data</span>, n <span class="op">=</span> <span class="fl">10L</span><span class="op">)</span>, 
    size <span class="op">=</span> <span class="fl">2</span>
  <span class="op">)</span> <span class="op">+</span> 
  <span class="fu">scale_fill_manual</span><span class="op">(</span>
    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"decreased"</span> <span class="op">=</span> <span class="st">"#cd5241"</span>, <span class="st">"increased"</span> <span class="op">=</span> <span class="st">"#207394"</span><span class="op">)</span>, 
    na.value <span class="op">=</span> <span class="st">"#cdcdcd"</span>
  <span class="op">)</span> <span class="op">+</span> 
  <span class="fu">labs</span><span class="op">(</span>
    x <span class="op">=</span> <span class="st">"log2(Fold-change)"</span>, 
    y <span class="op">=</span> <span class="st">"-log10(p-value)"</span>, 
    fill <span class="op">=</span> <span class="cn">NULL</span>, 
    caption <span class="op">=</span> <span class="st">"Labels indicate the 10 most significant marker-cluster pairs"</span>
  <span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-30-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="patient--and-sample-level-feature-extraction">Patient- and sample-level feature extraction<a class="anchor" aria-label="anchor" href="#patient--and-sample-level-feature-extraction"></a>
</h3>
<p>In addition to its functions for analyzing and visualizing CyTOF data at the single-cell and cluster levels, <code>{tidytof}</code>’s <code>tof_extract_features</code> verb allows users to aggregate single-cell and cluster-level information in order to summarize whole-samples (or whole-patients) from which cells were collected. These features can be useful for visualizing the differences between patients and samples in different experimental conditions or for building machine learning models.</p>
<p>To understand how the <code>tof_extract_features</code> verb works, it’s easiest to look at each of its subroutines (the members of the <code>tof_extract_*</code> function family) independently.</p>
<p>First, we have <code>tof_extract_proportion</code>, which extracts the proportion of cells in each cluster within each sample (with samples defined using the <code>group_cols</code> argument):</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># preprocess the numeric columns in the citrus dataset</span>
<span class="va">citrus_data</span> <span class="op">&lt;-</span> 
  <span class="va">citrus_data</span> <span class="op">%&gt;%</span> 
  <span class="fu">mutate</span><span class="op">(</span>cluster <span class="op">=</span> <span class="fu">str_c</span><span class="op">(</span><span class="st">"cluster"</span>, <span class="va">population_id</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tof_preprocess.html">tof_preprocess</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">citrus_data</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tof_extract_proportion.html">tof_extract_proportion</a></span><span class="op">(</span>
    cluster_col <span class="op">=</span> <span class="va">cluster</span>, 
    group_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">patient</span>, <span class="va">stimulation</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 × 10</span>
<span class="co">#&gt;   patient  stimulation `prop@cluster1` `prop@cluster2` `prop@cluster3`</span>
<span class="co">#&gt;   &lt;chr&gt;    &lt;chr&gt;                 &lt;dbl&gt;           &lt;dbl&gt;           &lt;dbl&gt;</span>
<span class="co">#&gt; 1 patient1 Basal                0.0190          0.0482           0.447</span>
<span class="co">#&gt; 2 patient1 BCR-XL               0.0109          0.0395           0.268</span>
<span class="co">#&gt; 3 patient2 Basal                0.0130          0.0280           0.491</span>
<span class="co">#&gt; 4 patient2 BCR-XL               0.0101          0.0143           0.358</span>
<span class="co">#&gt; 5 patient3 Basal                0.0326          0.0830           0.397</span>
<span class="co">#&gt; 6 patient3 BCR-XL               0.0200          0.0412           0.323</span>
<span class="co">#&gt; # … with 5 more variables: prop@cluster4 &lt;dbl&gt;, prop@cluster5 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   prop@cluster6 &lt;dbl&gt;, prop@cluster7 &lt;dbl&gt;, prop@cluster8 &lt;dbl&gt;</span></code></pre></div>
<p>Like all members of the <code>tof_extract_*</code> function family, <code>tof_extract_proportion</code> returns one row for each sample (defined as a unique combination of values of the <code>group_cols</code>) and one column for each extracted feature (above, one column for the proportion of each of the 8 clusters in <code>citrus_data</code>). These values can also be returned in “long” format by changing the <code>format</code> argument:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">citrus_data</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tof_extract_proportion.html">tof_extract_proportion</a></span><span class="op">(</span>
    cluster_col <span class="op">=</span> <span class="va">cluster</span>, 
    group_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">patient</span>, <span class="va">stimulation</span><span class="op">)</span>, 
    format <span class="op">=</span> <span class="st">"long"</span>
  <span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 × 4</span>
<span class="co">#&gt;   patient  stimulation cluster     prop</span>
<span class="co">#&gt;   &lt;chr&gt;    &lt;chr&gt;       &lt;chr&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt; 1 patient1 Basal       cluster1 0.0190 </span>
<span class="co">#&gt; 2 patient1 Basal       cluster2 0.0482 </span>
<span class="co">#&gt; 3 patient1 Basal       cluster3 0.447  </span>
<span class="co">#&gt; 4 patient1 Basal       cluster4 0.237  </span>
<span class="co">#&gt; 5 patient1 Basal       cluster5 0.00219</span>
<span class="co">#&gt; 6 patient1 Basal       cluster6 0.0759</span></code></pre></div>
<p>Another member of the same function family, <code>tof_extract_central_tendency</code>, computes the central tendency (e.g. mean or median) of user-specified markers in each cluster.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">citrus_data</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tof_extract_central_tendency.html">tof_extract_central_tendency</a></span><span class="op">(</span>
    cluster_col <span class="op">=</span> <span class="va">cluster</span>, 
    group_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">patient</span>, <span class="va">stimulation</span><span class="op">)</span>, 
    marker_cols <span class="op">=</span> <span class="fu">any_of</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD45_In115"</span>, <span class="st">"CD4_Nd145"</span>, <span class="st">"CD20_Sm147"</span><span class="op">)</span><span class="op">)</span>, 
    central_tendency_function <span class="op">=</span> <span class="va">mean</span>
  <span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 × 26</span>
<span class="co">#&gt;   patient  stimulation `CD45_In115@cluster1` `CD4_Nd145@clust… `CD20_Sm147@clus…</span>
<span class="co">#&gt;   &lt;chr&gt;    &lt;chr&gt;                       &lt;dbl&gt;             &lt;dbl&gt;             &lt;dbl&gt;</span>
<span class="co">#&gt; 1 patient1 Basal                        4.68            0.834               3.72</span>
<span class="co">#&gt; 2 patient1 BCR-XL                       4.80            0.186               4.11</span>
<span class="co">#&gt; 3 patient2 Basal                        4.89            0.879               3.55</span>
<span class="co">#&gt; 4 patient2 BCR-XL                       5.00            0.0379              3.91</span>
<span class="co">#&gt; 5 patient3 Basal                        4.98            0.818               3.60</span>
<span class="co">#&gt; 6 patient3 BCR-XL                       5.04            0.0516              3.92</span>
<span class="co">#&gt; # … with 21 more variables: CD45_In115@cluster2 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   CD4_Nd145@cluster2 &lt;dbl&gt;, CD20_Sm147@cluster2 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   CD45_In115@cluster3 &lt;dbl&gt;, CD4_Nd145@cluster3 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   CD20_Sm147@cluster3 &lt;dbl&gt;, CD45_In115@cluster4 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   CD4_Nd145@cluster4 &lt;dbl&gt;, CD20_Sm147@cluster4 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   CD45_In115@cluster5 &lt;dbl&gt;, CD4_Nd145@cluster5 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   CD20_Sm147@cluster5 &lt;dbl&gt;, CD45_In115@cluster6 &lt;dbl&gt;, …</span></code></pre></div>
<p><code>tof_extract_threshold</code> is similar to <code>tof_extract_central_tendency</code>, but calculates the proportion of cells above a user-specified expression value for each marker instead of a measure of central tendency:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">citrus_data</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tof_extract_threshold.html">tof_extract_threshold</a></span><span class="op">(</span>
    cluster_col <span class="op">=</span> <span class="va">cluster</span>, 
    group_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">patient</span>, <span class="va">stimulation</span><span class="op">)</span>, 
    marker_cols <span class="op">=</span> <span class="fu">any_of</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD45_In115"</span>, <span class="st">"CD4_Nd145"</span>, <span class="st">"CD20_Sm147"</span><span class="op">)</span><span class="op">)</span>, 
    threshold <span class="op">=</span> <span class="fl">5</span>
  <span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 × 26</span>
<span class="co">#&gt;   patient  stimulation `CD45_In115@cluste… `CD4_Nd145@cluste… `CD20_Sm147@clust…</span>
<span class="co">#&gt;   &lt;chr&gt;    &lt;chr&gt;                     &lt;dbl&gt;              &lt;dbl&gt;              &lt;dbl&gt;</span>
<span class="co">#&gt; 1 patient1 Basal                     0.365                  0             0.0769</span>
<span class="co">#&gt; 2 patient1 BCR-XL                    0.516                  0             0.0968</span>
<span class="co">#&gt; 3 patient2 Basal                     0.452                  0             0.0323</span>
<span class="co">#&gt; 4 patient2 BCR-XL                    0.554                  0             0.101 </span>
<span class="co">#&gt; 5 patient3 Basal                     0.549                  0             0.0552</span>
<span class="co">#&gt; 6 patient3 BCR-XL                    0.547                  0             0.0816</span>
<span class="co">#&gt; # … with 21 more variables: CD45_In115@cluster2_threshold &lt;dbl&gt;,</span>
<span class="co">#&gt; #   CD4_Nd145@cluster2_threshold &lt;dbl&gt;, CD20_Sm147@cluster2_threshold &lt;dbl&gt;,</span>
<span class="co">#&gt; #   CD45_In115@cluster3_threshold &lt;dbl&gt;, CD4_Nd145@cluster3_threshold &lt;dbl&gt;,</span>
<span class="co">#&gt; #   CD20_Sm147@cluster3_threshold &lt;dbl&gt;, CD45_In115@cluster4_threshold &lt;dbl&gt;,</span>
<span class="co">#&gt; #   CD4_Nd145@cluster4_threshold &lt;dbl&gt;, CD20_Sm147@cluster4_threshold &lt;dbl&gt;,</span>
<span class="co">#&gt; #   CD45_In115@cluster5_threshold &lt;dbl&gt;, CD4_Nd145@cluster5_threshold &lt;dbl&gt;,</span>
<span class="co">#&gt; #   CD20_Sm147@cluster5_threshold &lt;dbl&gt;, CD45_In115@cluster6_threshold &lt;dbl&gt;, …</span></code></pre></div>
<p>The two final members of the <code>tof_extract_*</code> function family – <code>tof_extract_emd</code> and <code>tof_extract_jsd</code> are designed specifically for comparing distributions of marker expression between stimulation conditions. As such, they must be given a <code>stimulation_col</code> that identifies which stimulation condition each cell is in, and a <code>basal_level</code> that specifies the reference (i.e. unstimulated) condition within the <code>stimulation_col</code>. With these additional arguments, <code>tof_extract_emd</code> computes the Earth-mover’s distance between each marker’s distribution in the stimulation conditions (within each cluster) and the basal condition; similarly, <code>tof_extract_jsd</code> computes the Jensen-Shannon divergence index between the same distributions. Both of these values are ways to compare how different 2 distributions are to one another and are more computationally expensive (but also higher-resolution) than simply comparing measures of central tendency.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Earth-mover's distance</span>
<span class="va">citrus_data</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tof_extract_emd.html">tof_extract_emd</a></span><span class="op">(</span>
    cluster_col <span class="op">=</span> <span class="va">cluster</span>, 
    group_cols <span class="op">=</span> <span class="va">patient</span>, 
    marker_cols <span class="op">=</span> <span class="fu">any_of</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD45_In115"</span>, <span class="st">"CD4_Nd145"</span>, <span class="st">"CD20_Sm147"</span><span class="op">)</span><span class="op">)</span>, 
    stimulation_col <span class="op">=</span> <span class="va">stimulation</span>, 
    basal_level <span class="op">=</span> <span class="st">"Basal"</span>
  <span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 × 25</span>
<span class="co">#&gt;   patient  `BCR-XL_CD45_In1… `BCR-XL_CD4_Nd14… `BCR-XL_CD20_Sm… `BCR-XL_CD45_In…</span>
<span class="co">#&gt;   &lt;chr&gt;                &lt;dbl&gt;             &lt;dbl&gt;            &lt;dbl&gt;            &lt;dbl&gt;</span>
<span class="co">#&gt; 1 patient1             0.870              2.48            12.5             1.45 </span>
<span class="co">#&gt; 2 patient2             1.11               7.02            10.9             0.730</span>
<span class="co">#&gt; 3 patient3             0.756              6.19            10.5             0.639</span>
<span class="co">#&gt; 4 patient4             2.64               6.78             9.47            3.27 </span>
<span class="co">#&gt; 5 patient5             0.593              7.49             7.70            0.781</span>
<span class="co">#&gt; 6 patient6             0.656              4.74             8.72            1.57 </span>
<span class="co">#&gt; # … with 20 more variables: BCR-XL_CD4_Nd145@cluster7_emd &lt;dbl&gt;,</span>
<span class="co">#&gt; #   BCR-XL_CD20_Sm147@cluster7_emd &lt;dbl&gt;, BCR-XL_CD45_In115@cluster4_emd &lt;dbl&gt;,</span>
<span class="co">#&gt; #   BCR-XL_CD4_Nd145@cluster4_emd &lt;dbl&gt;, BCR-XL_CD20_Sm147@cluster4_emd &lt;dbl&gt;,</span>
<span class="co">#&gt; #   BCR-XL_CD45_In115@cluster2_emd &lt;dbl&gt;, BCR-XL_CD4_Nd145@cluster2_emd &lt;dbl&gt;,</span>
<span class="co">#&gt; #   BCR-XL_CD20_Sm147@cluster2_emd &lt;dbl&gt;, BCR-XL_CD45_In115@cluster6_emd &lt;dbl&gt;,</span>
<span class="co">#&gt; #   BCR-XL_CD4_Nd145@cluster6_emd &lt;dbl&gt;, BCR-XL_CD20_Sm147@cluster6_emd &lt;dbl&gt;,</span>
<span class="co">#&gt; #   BCR-XL_CD45_In115@cluster8_emd &lt;dbl&gt;, …</span></code></pre></div>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Jensen-Shannon Divergence</span>
<span class="va">citrus_data</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tof_extract_jsd.html">tof_extract_jsd</a></span><span class="op">(</span>
    cluster_col <span class="op">=</span> <span class="va">cluster</span>, 
    group_cols <span class="op">=</span> <span class="va">patient</span>,  
    marker_cols <span class="op">=</span> <span class="fu">any_of</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD45_In115"</span>, <span class="st">"CD4_Nd145"</span>, <span class="st">"CD20_Sm147"</span><span class="op">)</span><span class="op">)</span>, 
    stimulation_col <span class="op">=</span> <span class="va">stimulation</span>, 
    basal_level <span class="op">=</span> <span class="st">"Basal"</span>
  <span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 × 25</span>
<span class="co">#&gt;   patient  `BCR-XL_CD45_In1… `BCR-XL_CD4_Nd14… `BCR-XL_CD20_Sm… `BCR-XL_CD45_In…</span>
<span class="co">#&gt;   &lt;chr&gt;                &lt;dbl&gt;             &lt;dbl&gt;            &lt;dbl&gt;            &lt;dbl&gt;</span>
<span class="co">#&gt; 1 patient1           0.0331             0.0479            0.348           0.0580</span>
<span class="co">#&gt; 2 patient2           0.00805            0.168             0.402           0.0159</span>
<span class="co">#&gt; 3 patient3           0.0114             0.113             0.358           0.0205</span>
<span class="co">#&gt; 4 patient4           0.0303             0.136             0.205           0.0390</span>
<span class="co">#&gt; 5 patient5           0.00908            0.0782            0.283           0.0261</span>
<span class="co">#&gt; 6 patient6           0.00919            0.0341            0.222           0.0450</span>
<span class="co">#&gt; # … with 20 more variables: BCR-XL_CD4_Nd145@cluster7_jsd &lt;dbl&gt;,</span>
<span class="co">#&gt; #   BCR-XL_CD20_Sm147@cluster7_jsd &lt;dbl&gt;, BCR-XL_CD45_In115@cluster4_jsd &lt;dbl&gt;,</span>
<span class="co">#&gt; #   BCR-XL_CD4_Nd145@cluster4_jsd &lt;dbl&gt;, BCR-XL_CD20_Sm147@cluster4_jsd &lt;dbl&gt;,</span>
<span class="co">#&gt; #   BCR-XL_CD45_In115@cluster2_jsd &lt;dbl&gt;, BCR-XL_CD4_Nd145@cluster2_jsd &lt;dbl&gt;,</span>
<span class="co">#&gt; #   BCR-XL_CD20_Sm147@cluster2_jsd &lt;dbl&gt;, BCR-XL_CD45_In115@cluster6_jsd &lt;dbl&gt;,</span>
<span class="co">#&gt; #   BCR-XL_CD4_Nd145@cluster6_jsd &lt;dbl&gt;, BCR-XL_CD20_Sm147@cluster6_jsd &lt;dbl&gt;,</span>
<span class="co">#&gt; #   BCR-XL_CD45_In115@cluster8_jsd &lt;dbl&gt;, …</span></code></pre></div>
<p>Finally, the <code>tof_extract_features</code> verb provides a wrapper to each of the members of its function family, allowing users to extract multiple features types at once. For example, the following code extracts the proportion of each cluster, median of several markers in each cluster, and EMD between the basal condition and stimulated condition in each cluster for all patients in <code>citrus_data</code>.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">citrus_data</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tof_extract_features.html">tof_extract_features</a></span><span class="op">(</span>
    cluster_col <span class="op">=</span> <span class="va">cluster</span>, 
    group_cols <span class="op">=</span> <span class="va">patient</span>, 
    stimulation_col <span class="op">=</span> <span class="va">stimulation</span>,
    lineage_cols <span class="op">=</span> <span class="fu">any_of</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD45_In115"</span>, <span class="st">"CD20_Sm147"</span>, <span class="st">"CD33_Nd148"</span><span class="op">)</span><span class="op">)</span>, 
    signaling_cols <span class="op">=</span> <span class="fu">any_of</span><span class="op">(</span><span class="va">signaling_markers</span><span class="op">)</span>, 
    signaling_method <span class="op">=</span> <span class="st">"emd"</span>, 
    basal_level <span class="op">=</span> <span class="st">"Basal"</span>
  <span class="op">)</span> </code></pre></div>
</div>
<div class="section level3">
<h3 id="outcomes-modeling">Outcomes modeling<a class="anchor" aria-label="anchor" href="#outcomes-modeling"></a>
</h3>
<p>[Under construction]</p>
</div>
<div class="section level3">
<h3 id="post-processing">Post-processing<a class="anchor" aria-label="anchor" href="#post-processing"></a>
</h3>
<p>[Under construction]</p>
</div>
<div class="section level3">
<h3 id="writing-data">Writing data<a class="anchor" aria-label="anchor" href="#writing-data"></a>
</h3>
<p>[Under construction]</p>
</div>
</div>
<div class="section level2">
<h2 id="tidytofs-design-principles-and-tips">
<code>{tidytof}</code>’s Design Principles (and Tips)<a class="anchor" aria-label="anchor" href="#tidytofs-design-principles-and-tips"></a>
</h2>
<p>[Under construction]</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing tidytof</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Timothy Keyes <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/keyes-timothy/tidytof/actions" class="external-link"><img src="https://github.com/keyes-timothy/tidytof/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://app.codecov.io/gh/keyes-timothy/tidytof?branch=main" class="external-link"><img src="https://codecov.io/gh/keyes-timothy/tidytof/branch/main/graph/badge.svg" alt="Codecov test coverage"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Timothy Keyes.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
