<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Read this vignette if this is your first time using tidytof. You'll  learn about tidytof's basic design schema, its main capabilities,  and where to find more information. 
">
<title>Getting started with tidytof • tidytof</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Getting started with tidytof">
<meta property="og:description" content="Read this vignette if this is your first time using tidytof. You'll  learn about tidytof's basic design schema, its main capabilities,  and where to find more information. 
">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">tidytof</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/tidytof.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Extending {tidytof}</h6>
    <a class="dropdown-item" href="../articles/contributing-to-tidytof.html">How to contribute code</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Cell-level operations</h6>
    <a class="dropdown-item" href="../articles/reading-and-writing-data.html">Reading and writing data</a>
    <a class="dropdown-item" href="../articles/quality-control.html">Quality control</a>
    <a class="dropdown-item" href="../articles/preprocessing.html">Preprocessing</a>
    <a class="dropdown-item" href="../articles/downsampling.html">Downsampling</a>
    <a class="dropdown-item" href="../articles/dimensionality-reduction.html">Dimensionality reduction</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Cluster-level operations</h6>
    <a class="dropdown-item" href="../articles/clustering.html">Clustering and metaclustering</a>
    <a class="dropdown-item" href="../articles/differential-discovery-analysis.html">Differential discovery analysis</a>
    <a class="dropdown-item" href="../articles/feature-extraction.html">Feature extraction</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Sample-level operations</h6>
    <a class="dropdown-item" href="../articles/modeling.html">Building predictive models</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Getting started with tidytof</h1>
                        <h4 data-toc-skip class="author">Timothy
Keyes</h4>
            
            <h4 data-toc-skip class="date">2024-03-25</h4>
      
      
      <div class="d-none name"><code>tidytof.Rmd</code></div>
    </div>

    
    
<p>Analyzing single-cell data can be surprisingly complicated. This is
partially because single-cell data analysis is an incredibly active area
of research, with new methods being published on a weekly - or even
daily! - basis. Accordingly, when new tools are published, they often
require researchers to learn unique, method-specific application
programming interfaces (APIs) with distinct requirements for input data
formatting, function syntax, and output data structure. On the other
hand, analyzing single-cell data can be challenging because it often
involves simultaneously asking questions at multiple levels of
biological scope - the single-cell level, the cell subpopulation
(i.e. cluster) level, and the whole-sample or whole-patient level - each
of which has distinct data processing needs.</p>
<p>To address both of these challenges for high-dimensional cytometry,
<a href="https://keyes-timothy.github.io/tidytof">tidytof</a> implements a concise, integrated “grammar” of
single-cell data analysis capable of answering a variety of biological
questions. Available as an open-source R package, <a href="https://keyes-timothy.github.io/tidytof">tidytof</a>
provides an easy-to-use pipeline for analyzing high-dimensional
cytometry data by automating many common data-processing tasks under a
common <a href="https://r4ds.had.co.nz/tidy-data.html" class="external-link">“tidy data”</a>
interface. This vignette introduces you to the tidytof’s high-level API
and shows quick examples of how they can be applied to high-dimensional
cytometry datasets.</p>
<div class="section level3">
<h3 id="prerequisites">Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h3>
<p><a href="https://keyes-timothy.github.io/tidytof">tidytof</a> makes heavy use of two concepts that may be
unfamiliar to R beginners. The first is the pipe (<code>|&gt;</code>),
which you can read about <a href="https://r4ds.had.co.nz/pipes.html" class="external-link">here</a>. The second is
“grouping” data in a <code>data.frame</code> or <code>tibble</code>
using <code><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">dplyr::group_by</a></code>, which you can read about <a href="https://dplyr.tidyverse.org/articles/grouping.html" class="external-link">here</a>. Most
<a href="https://keyes-timothy.github.io/tidytof">tidytof</a> users will also benefit from a relatively
in-depth understanding of the dplyr package, which has a wonderful
introductory vignette here:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/vignette.html" class="external-link">vignette</a></span><span class="op">(</span><span class="st">"dplyr"</span><span class="op">)</span></span></code></pre></div>
<p>Everything else should be self-explanatory for both beginner and
advanced R users, though if you have <em>zero</em> background in running
R code, you should read <a href="https://r4ds.had.co.nz/workflow-basics.html" class="external-link">this chapter</a> of
<a href="https://r4ds.had.co.nz/index.html" class="external-link">R for Data Science</a> by
Hadley Wickham.</p>
</div>
<div class="section level3">
<h3 id="workflow-basics">Workflow basics<a class="anchor" aria-label="anchor" href="#workflow-basics"></a>
</h3>
<p>Broadly speaking, <a href="https://keyes-timothy.github.io/tidytof">tidytof</a>’s functionality is organized
to support the 3 levels of analysis inherent to single-cell data
described above:</p>
<ol style="list-style-type: decimal">
<li>Reading, writing, preprocessing, and visualizing data at the level
of <strong>individual cells</strong>
</li>
<li>Identifying and describing cell <strong>subpopulations</strong> or
<strong>clusters</strong>
</li>
<li>Building models (for inference or prediction) at the level of
<strong>patients</strong> or <strong>samples</strong>
</li>
</ol>
<p><a href="https://keyes-timothy.github.io/tidytof">tidytof</a> provides functions (or “verbs”) that operate
at each of these levels of analysis:</p>
<ul>
<li>
<p>Cell-level data:</p>
<ul>
<li>
<code><a href="../reference/tof_read_data.html">tof_read_data()</a></code> reads single-cell data from FCS or CSV
files on disk into a tidy data frame called a <code>tof_tbl</code>.
<code>tof_tbl</code>s represent each cell as a row and each protein
measurement (or other piece of information associated with a given cell)
as a column.</li>
<li>
<code><a href="../reference/tof_preprocess.html">tof_preprocess()</a></code> transforms protein expression values
using a user-provided function (i.e. log-transformation, centering,
scaling)</li>
<li>
<code><a href="../reference/tof_downsample.html">tof_downsample()</a></code> reduces the number of cells in a
<code>tof_tibble</code> via subsampling.</li>
<li>
<code><a href="../reference/tof_reduce_dimensions.html">tof_reduce_dimensions()</a></code> performs dimensionality
reduction (across columns)</li>
<li>
<code>tof_write_data</code> writes single-cell data in a
<code>tof_tibble</code> back to disk in the form of an FCS or CSV
file.</li>
</ul>
</li>
<li>
<p>Cluster-level data:</p>
<ul>
<li>
<code><a href="../reference/tof_cluster.html">tof_cluster()</a></code> clusters cells using one of several
algorithms commonly applied to high-dimensional cytometry data</li>
<li>
<code><a href="../reference/tof_metacluster.html">tof_metacluster()</a></code> agglomerates clusters into a smaller
number of metaclusters</li>
<li>
<code><a href="../reference/tof_analyze_abundance.html">tof_analyze_abundance()</a></code> performs differential abundance
analysis (DAA) for clusters or metaclusters across experimental
groups</li>
<li>
<code><a href="../reference/tof_analyze_expression.html">tof_analyze_expression()</a></code> performs differential
expression analysis (DEA) for clusters’ or metaclusters’ marker
expression levels across experimental groups</li>
<li>
<code><a href="../reference/tof_extract_features.html">tof_extract_features()</a></code> computes summary statistics
(such as mean marker expression) for each cluster. Also (optionally)
pivots these summary statistics into a sample-level tidy data frame in
which each row represents a sample and each column represents a
cluster-level summary statistic.</li>
</ul>
</li>
<li>
<p>Sample-level data:</p>
<ul>
<li>
<code><a href="../reference/tof_split_data.html">tof_split_data()</a></code> splits sample-level data into a
training and test set for predictive modeling</li>
<li>
<code><a href="../reference/tof_create_grid.html">tof_create_grid()</a></code> creates an elastic net hyperparameter
search grid for model tuning</li>
<li>
<code><a href="../reference/tof_train_model.html">tof_train_model()</a></code> trains a sample-level elastic net
model and saves it as a <code>tof_model</code> object</li>
<li>
<code><a href="../reference/tof_predict.html">tof_predict()</a></code> Applies a trained <code>tof_model</code>
to new data to predict sample-level outcomes</li>
<li>
<code><a href="../reference/tof_assess_model.html">tof_assess_model()</a></code> calculates performance metrics for a
trained <code>tof_model</code>
</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="tidytof-verb-syntax">{tidytof} verb syntax<a class="anchor" aria-label="anchor" href="#tidytof-verb-syntax"></a>
</h3>
<p>With very few exceptions, <a href="https://keyes-timothy.github.io/tidytof">tidytof</a> functions follow a
specific, shared syntax that involves 3 types of arguments that always
occur in the same order. These argument types are as follows:</p>
<ol style="list-style-type: decimal">
<li>For almost all <a href="https://keyes-timothy.github.io/tidytof">tidytof</a> functions, the first argument
is a data frame (or tibble). This enables the use of the pipe
(<code>|&gt;</code>) for multi-step calculations, which means that your
first argument for most functions will be implicit (passed from the
previous function using the pipe). This also means that most
<a href="https://keyes-timothy.github.io/tidytof">tidytof</a> functions are so-called <a href="https://cran.r-project.org/web/packages/dplyr/vignettes/two-table.html" class="external-link">“single-table
verbs,”</a> with the exception of <code>tof_cluster_ddpr</code>, which
is a “two-table verb” (for details about how to use
<code>tof_cluster_ddpr</code>, see the “clustering-and-metaclustering”
vignette).</li>
<li>The second group of arguments are called <em>column
specifications</em>, and they end in the suffix <code>_col</code> or
<code>_cols</code>. Column specifications are unquoted column names that
tell a <a href="https://keyes-timothy.github.io/tidytof">tidytof</a> verb which columns to compute over for a
particular operation. For example, the <code>cluster_cols</code>
argument in <code>tof_cluster</code> allows the user to specify which
column in the input data frames should be used to perform the
clustering. Regardless of which verb requires them, column
specifications support <a href="https://tidyselect.r-lib.org/reference/language.html" class="external-link">tidyselect
helpers</a> and follow the same rules for tidyselection as tidyverse
verbs like <code><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">dplyr::select()</a></code> and
<code><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">tidyr::pivot_longer()</a></code>.</li>
<li>Finally, the third group of arguments for each
<a href="https://keyes-timothy.github.io/tidytof">tidytof</a> verb are called <em>method specifications</em>,
and they’re comprised of every argument that isn’t an input data frame
or a column specification. Whereas column specifications represent which
columns should be used to perform an operation, method specifications
represent the details of how that operation should be performed. For
example, the <code><a href="../reference/tof_cluster_phenograph.html">tof_cluster_phenograph()</a></code> function requires the
method specification <code>num_neighbors</code>, which specifies how
many nearest neighbors should be used to construct the PhenoGraph
algorithm’s k-nearest-neighbor graph. In most cases,
<a href="https://keyes-timothy.github.io/tidytof">tidytof</a> sets reasonable defaults for each verb’s
particular method specifications, but your workflows are can also be
customized by experimenting with non-default values.</li>
</ol>
<p>The following code demonstrates how <a href="https://keyes-timothy.github.io/tidytof">tidytof</a> verb
syntax looks in practice, with column and method specifications
explicitly pointed out:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">ddpr_data</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">777L</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ddpr_data</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/tof_preprocess.html">tof_preprocess</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/tof_cluster.html">tof_cluster</a></span><span class="op">(</span></span>
<span>        cluster_cols <span class="op">=</span> <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">starts_with</a></span><span class="op">(</span><span class="st">"cd"</span><span class="op">)</span>, <span class="co"># column specification</span></span>
<span>        method <span class="op">=</span> <span class="st">"phenograph"</span>, <span class="co"># method specification,</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/tof_metacluster.html">tof_metacluster</a></span><span class="op">(</span></span>
<span>        cluster_col <span class="op">=</span> <span class="va">.phenograph_cluster</span>, <span class="co"># column specification</span></span>
<span>        num_metaclusters <span class="op">=</span> <span class="fl">4</span>, <span class="co"># method specification</span></span>
<span>        method <span class="op">=</span> <span class="st">"kmeans"</span> <span class="co"># method specification</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/tof_downsample.html">tof_downsample</a></span><span class="op">(</span></span>
<span>        group_cols <span class="op">=</span> <span class="va">.kmeans_metacluster</span>, <span class="co"># column specification</span></span>
<span>        num_cells <span class="op">=</span> <span class="fl">200</span>, <span class="co"># method specification</span></span>
<span>        method <span class="op">=</span> <span class="st">"constant"</span> <span class="co"># method specification</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="../reference/tof_plot_cells_layout.html">tof_plot_cells_layout</a></span><span class="op">(</span></span>
<span>        knn_cols <span class="op">=</span> <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">starts_with</a></span><span class="op">(</span><span class="st">"cd"</span><span class="op">)</span>, <span class="co"># column specification</span></span>
<span>        color_col <span class="op">=</span> <span class="va">.kmeans_metacluster</span>, <span class="co"># column specification</span></span>
<span>        num_neighbors <span class="op">=</span> <span class="fl">7L</span>, <span class="co"># method specification</span></span>
<span>        node_size <span class="op">=</span> <span class="fl">2L</span> <span class="co"># method specification</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p><img src="tidytof_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="pipelines">Pipelines<a class="anchor" aria-label="anchor" href="#pipelines"></a>
</h3>
<p><a href="https://keyes-timothy.github.io/tidytof">tidytof</a> verbs can be used on their own or in
combination with one another using the pipe (<code>|&gt;</code>)
operator. For example, here is a multistep “pipeline” that takes a
built-in <a href="https://keyes-timothy.github.io/tidytof">tidytof</a> dataset and performs the following
analytical steps:</p>
<ol style="list-style-type: decimal">
<li><p>Arcsinh-transform each column of protein measurements (the
default behavior of the <code>tof_preprocess</code> verb</p></li>
<li><p>Cluster our cells based on the surface markers in our
panel</p></li>
<li><p>Downsample the dataset such that 100 random cells are picked from
each cluster</p></li>
<li><p>Perform dimensionality reduction on the downsampled dataset using
tSNE</p></li>
<li><p>Visualize the clusters using a low-dimensional tSNE
embedding</p></li>
</ol>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ddpr_data</span> <span class="op">|&gt;</span></span>
<span>    <span class="co"># step 1</span></span>
<span>    <span class="fu"><a href="../reference/tof_preprocess.html">tof_preprocess</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="co"># step 2</span></span>
<span>    <span class="fu"><a href="../reference/tof_cluster.html">tof_cluster</a></span><span class="op">(</span></span>
<span>        cluster_cols <span class="op">=</span> <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">starts_with</a></span><span class="op">(</span><span class="st">"cd"</span><span class="op">)</span>,</span>
<span>        method <span class="op">=</span> <span class="st">"phenograph"</span>,</span>
<span>        <span class="co"># num_metaclusters = 4L,</span></span>
<span>        seed <span class="op">=</span> <span class="fl">2020L</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="co"># step 3</span></span>
<span>    <span class="fu"><a href="../reference/tof_downsample.html">tof_downsample</a></span><span class="op">(</span></span>
<span>        group_cols <span class="op">=</span> <span class="va">.phenograph_cluster</span>,</span>
<span>        method <span class="op">=</span> <span class="st">"constant"</span>,</span>
<span>        num_cells <span class="op">=</span> <span class="fl">400</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="co"># step 4</span></span>
<span>    <span class="fu"><a href="../reference/tof_reduce_dimensions.html">tof_reduce_dimensions</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"tsne"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="co"># step 5</span></span>
<span>    <span class="fu"><a href="../reference/tof_plot_cells_embedding.html">tof_plot_cells_embedding</a></span><span class="op">(</span></span>
<span>        embedding_cols <span class="op">=</span> <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">contains</a></span><span class="op">(</span><span class="st">"tsne"</span><span class="op">)</span>,</span>
<span>        color_col <span class="op">=</span> <span class="va">.phenograph_cluster</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required namespace: Rtsne</span></span></code></pre></div>
<p><img src="tidytof_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="other-tips">Other tips<a class="anchor" aria-label="anchor" href="#other-tips"></a>
</h3>
<p><a href="https://keyes-timothy.github.io/tidytof">tidytof</a> was designed by a multidisciplinary team of
wet-lab biologists, bioinformaticians, and physician-scientists who
analyze high-dimensional cytometry and other kinds of single-cell data
to solve a variety of problems. As a result, <a href="https://keyes-timothy.github.io/tidytof">tidytof</a>’s
high-level API was designed with great care to mirror that of the
<a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a> itself - that is, to be <a href="https://design.tidyverse.org/unifying-principles.html" class="external-link">human-centered,
consistent, composable, and inclusive</a> for a wide userbase.</p>
<p>Practically speaking, this means a few things about using
<a href="https://keyes-timothy.github.io/tidytof">tidytof</a>.</p>
<p>First, it means that <a href="https://keyes-timothy.github.io/tidytof">tidytof</a> was designed with a few
quality-of-life features in mind. For example, you may notice that most
<a href="https://keyes-timothy.github.io/tidytof">tidytof</a> functions begin with the prefix
<code>tof_</code>. This is intentional, as it will allow you to use your
development environment’s code-completing software to search for
<a href="https://keyes-timothy.github.io/tidytof">tidytof</a> functions easily (even if you can’t remember a
specific function name). For this reason, we recommend using
<a href="https://keyes-timothy.github.io/tidytof">tidytof</a> within the RStudio development environment;
however, many code editors have predictive text functionality that
serves a similar function. In general, <a href="https://keyes-timothy.github.io/tidytof">tidytof</a> verbs are
organized in such a way that your IDE’s code-completion tools should
also allow you to search for (and compare) related functions with
relative ease. (For instance, the <code>tof_cluster_</code> prefix is
used for all clustering functions, and the <code>tof_downsample_</code>
prefix is used for all downsampling functions).</p>
<p>Second, it means that <a href="https://keyes-timothy.github.io/tidytof">tidytof</a> functions
<em>should</em> be relatively intuitive to use due to their shared logic
- in other words, if you understand how to use one
<a href="https://keyes-timothy.github.io/tidytof">tidytof</a> function, you should understand how to use most
of the others. An example of shared logic across <a href="https://keyes-timothy.github.io/tidytof">tidytof</a>
functions is the argument <code>group_cols</code>, which shows up in
multiple verbs (<code>tof_downsample</code>, <code>tof_cluster</code>,
<code>tof_daa</code>, <code>tof_dea</code>,
<code>tof_extract_features</code>, and <code>tof_write_data</code>). In
each case, <code>group_cols</code> works the same way: it accepts an
unquoted vector of column names (specified manually or using <a href="https://r4ds.had.co.nz/transform.html#select" class="external-link">tidyselection</a>)
that should be used to group cells before an operation is performed.
This idea generalizes throughout <a href="https://keyes-timothy.github.io/tidytof">tidytof</a>: if you see an
argument in one place, it will behave identically (or at least very
similarly) wherever else you encounter it.</p>
<p>Finally, it means that <a href="https://keyes-timothy.github.io/tidytof">tidytof</a> is optimized first for
ease-of-use, then for performance. Because humans and computers interact
with data differently, there is always a trade-off between choosing a
data representation that is intuitive to a human user vs. choosing a
data representation optimized for computational speed and memory
efficiency. When these design choices conflict with one another, our
team tends to err on the side of choosing a representation that is
easy-to-understand for users even at the expense of small performance
costs. Ultimately, this means that <a href="https://keyes-timothy.github.io/tidytof">tidytof</a> may not be the
optimal tool for every high-dimensional cytometry analysis, though
hopefully its general framework will provide most users with some useful
functionality.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.3 (2024-02-29)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] tidytof_0.99.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] gridExtra_2.3       rlang_1.1.3         magrittr_2.0.3     </span></span>
<span><span class="co">#&gt;   [4] matrixStats_1.2.0   compiler_4.3.3      systemfonts_1.0.6  </span></span>
<span><span class="co">#&gt;   [7] vctrs_0.6.5         stringr_1.5.1       pkgconfig_2.0.3    </span></span>
<span><span class="co">#&gt;  [10] shape_1.4.6.1       fastmap_1.1.1       labeling_0.4.3     </span></span>
<span><span class="co">#&gt;  [13] ggraph_2.2.1        utf8_1.2.4          rmarkdown_2.26     </span></span>
<span><span class="co">#&gt;  [16] prodlim_2023.08.28  tzdb_0.4.0          ragg_1.3.0         </span></span>
<span><span class="co">#&gt;  [19] purrr_1.0.2         xfun_0.42           glmnet_4.1-8       </span></span>
<span><span class="co">#&gt;  [22] cachem_1.0.8        jsonlite_1.8.8      recipes_1.0.10     </span></span>
<span><span class="co">#&gt;  [25] highr_0.10          tweenr_2.0.3        parallel_4.3.3     </span></span>
<span><span class="co">#&gt;  [28] R6_2.5.1            bslib_0.6.2         stringi_1.8.3      </span></span>
<span><span class="co">#&gt;  [31] parallelly_1.37.1   rpart_4.1.23        lubridate_1.9.3    </span></span>
<span><span class="co">#&gt;  [34] jquerylib_0.1.4     Rcpp_1.0.12         iterators_1.0.14   </span></span>
<span><span class="co">#&gt;  [37] knitr_1.45          future.apply_1.11.1 readr_2.1.5        </span></span>
<span><span class="co">#&gt;  [40] flowCore_2.14.2     Matrix_1.6-5        splines_4.3.3      </span></span>
<span><span class="co">#&gt;  [43] nnet_7.3-19         igraph_2.0.3        timechange_0.3.0   </span></span>
<span><span class="co">#&gt;  [46] tidyselect_1.2.1    yaml_2.3.8          viridis_0.6.5      </span></span>
<span><span class="co">#&gt;  [49] timeDate_4032.109   doParallel_1.0.17   codetools_0.2-19   </span></span>
<span><span class="co">#&gt;  [52] listenv_0.9.1       lattice_0.22-5      tibble_3.2.1       </span></span>
<span><span class="co">#&gt;  [55] Biobase_2.62.0      withr_3.0.0         Rtsne_0.17         </span></span>
<span><span class="co">#&gt;  [58] evaluate_0.23       future_1.33.1       desc_1.4.3         </span></span>
<span><span class="co">#&gt;  [61] survival_3.5-8      polyclip_1.10-6     pillar_1.9.0       </span></span>
<span><span class="co">#&gt;  [64] foreach_1.5.2       stats4_4.3.3        generics_0.1.3     </span></span>
<span><span class="co">#&gt;  [67] RcppHNSW_0.6.0      S4Vectors_0.40.2    hms_1.1.3          </span></span>
<span><span class="co">#&gt;  [70] ggplot2_3.5.0       munsell_0.5.0       scales_1.3.0       </span></span>
<span><span class="co">#&gt;  [73] globals_0.16.3      class_7.3-22        glue_1.7.0         </span></span>
<span><span class="co">#&gt;  [76] tools_4.3.3         data.table_1.15.2   gower_1.0.1        </span></span>
<span><span class="co">#&gt;  [79] fs_1.6.3            graphlayouts_1.1.1  tidygraph_1.3.1    </span></span>
<span><span class="co">#&gt;  [82] grid_4.3.3          yardstick_1.3.1     tidyr_1.3.1        </span></span>
<span><span class="co">#&gt;  [85] RProtoBufLib_2.14.1 ipred_0.9-14        colorspace_2.1-0   </span></span>
<span><span class="co">#&gt;  [88] ggforce_0.4.2       cli_3.6.2           textshaping_0.3.7  </span></span>
<span><span class="co">#&gt;  [91] fansi_1.0.6         cytolib_2.14.1      viridisLite_0.4.2  </span></span>
<span><span class="co">#&gt;  [94] lava_1.8.0          dplyr_1.1.4         gtable_0.3.4       </span></span>
<span><span class="co">#&gt;  [97] sass_0.4.9          digest_0.6.35       BiocGenerics_0.48.1</span></span>
<span><span class="co">#&gt; [100] ggrepel_0.9.5       farver_2.1.1        memoise_2.0.1      </span></span>
<span><span class="co">#&gt; [103] htmltools_0.5.7     pkgdown_2.0.7       lifecycle_1.0.4    </span></span>
<span><span class="co">#&gt; [106] hardhat_1.3.1       MASS_7.3-60.0.1</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Timothy Keyes.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
